src/helpers/StateTree.js

```diff
@@ -0,0 +1,62 @@
+
+/**
+ * StateTree helper class to simplify receiving state changed updates for nested onChange handlers.
+ */
+export class StateTree {
+	/**
+	 * Creates a new StateTree.
+	 * @param {Function} onStateChanged A state change handler that is called whenever the state of the tree changes by adding, replacing, or removing items.
+	 * @param {*} [item] The item to be stored at the root of the tree.
+	 */
+	constructor(onStateChanged, item, root) {
+		this._children = [];
+		this._onStateChanged = onStateChanged;
+		this._item = item;
+		this._root = root ? root : this;
+	}
+
+	/**
+	 * Attaches or overwrites a child node on the tree.
+	 * @param {Number} index  The location the child will be attached on the parent node.
+	 * @param {*} [item] The item to be stored in the newly created child node.
+	 * @returns {StateTree} The child node that was attached to the tree.
+	 */
+	setChild(index, item) {
+		const child = new StateTree(this._onStateChanged, item, this._root);
+		this._children[index] = child;
+
+		this._notifyStateChanged(true);
+
+		return child;
+	}
+
+	/**
+	 * Removes all attached child nodes.
+	 */
+	removeAllChildren() {
+		const hasStateChanged = this._children.length > 0;
+
+		this._children = [];
+
+		this._notifyStateChanged(hasStateChanged);
+	}
+
+	/**
+	 * Retrieves all items stored in the tree without guaranteeing any particular order.
+	 * @returns {Array} All items stored in the tree.
+	 */
+	items() {
+		let items = this._item ? [this._item] : [];
+		this._children.forEach(child => {
+			items = items.concat(child.items());
+		});
+		return items;
+	}
+
+	_notifyStateChanged(hasStateChanged) {
+		if (!hasStateChanged) {
+			return;
+		}
+		this._onStateChanged(this._root);
```

Comment:
What's your use case that your root entity cares about children changing?  My assumption is embedded representation siren entities.  If each child has a call back available for updating your ui state, updating the root would cause a fuller reflow of all the callbacks right?

Like a typical use case for many nested children ends up looking like:

```
consotriumTokenCollection.consortiumTokenEntities((consortiumEntity) => {
	consortiumEntity.rootOrganizationEntity((rootEntity) => {
         	rootEntity.organization((orgEntity) => {
                        //handler stuff that would get reinited
                }
         }
}
```